packages/server/lib/StudioLifecycleManager.ts

```diff
@@ -0,0 +1,143 @@
+  /**
+   * Initialize the studio manager and possibly set up protocol.
+   * Also registers this instance in the data context.
+   * @param projectId The project ID
+   * @param cloudDataSource The cloud data source
+   * @param cfg The project configuration
+   * @param debugData Debug data for the configuration
+   * @param ctx Data context to register this instance with
+   */
+  initializeStudioManager ({
+    projectId,
+    cloudDataSource,
+    cfg,
+    debugData,
+    ctx,
+  }: {
+    projectId?: string
+    cloudDataSource: CloudDataSource
+    cfg: Cfg
+    debugData: any
+    ctx: DataContext
+  }): void {
+    debug('Initializing studio manager')
+
+    const studioManagerPromise = getAndInitializeStudioManager({
+      projectId,
+      cloudDataSource,
+    }).then(async (studioManager) => {
+      if (studioManager.status === 'ENABLED') {
+        debug('Cloud studio is enabled - setting up protocol')
+        const protocolManager = new ProtocolManager()
+        const protocolUrl = routes.apiRoutes.captureProtocolCurrent()
+        const script = await api.getCaptureProtocolScript(protocolUrl)
+
+        await protocolManager.prepareProtocol(script, {
+          runId: 'studio',
+          projectId: cfg.projectId,
+          testingType: cfg.testingType,
+          cloudApi: {
+            url: routes.apiUrl,
+            retryWithBackoff: api.retryWithBackoff,
+            requestPromise: api.rp,
+          },
+          projectConfig: _.pick(cfg, ['devServerPublicPathRoute', 'port', 'proxyUrl', 'namespace']),
+          mountVersion: api.runnerCapabilities.protocolMountVersion,
+          debugData,
+          mode: 'studio',
+        })
+
+        studioManager.protocolManager = protocolManager
+        studioManager.isProtocolEnabled = true
+      } else {
+        debug('Cloud studio is not enabled - skipping protocol setup')
+      }
+
+      return studioManager
+    }).catch((err) => {
+      debug('Error during studio manager setup: %o', err)
+
+      return null
+    })
+
+    this.studioManagerPromise = studioManagerPromise
+
+    // When the promise resolves, call all the listeners
+    void studioManagerPromise.then(() => {
+      debug('Studio is ready')
+      this.studioReady = true
+      this.callRegisteredListeners()
+    })
+
+    // Register this instance in the data context
+    ctx.update((data) => {
+      data.studioLifecycleManager = this
+    })
+  }
+
+  isStudioReady (): boolean {
+    return this.studioReady
+  }
+
+  async getStudio () {
+    if (!this.studioManagerPromise) {
+      throw new Error('Studio manager has not been initialized')
+    }
+
+    return await this.studioManagerPromise
+  }
+
+  private callRegisteredListeners (listener?: ((studioManager: StudioManager) => void)) {
+    if (!this.studioManagerPromise || !this.studioReady) {
+      throw new Error('Studio manager has not been initialized')
+    }
+
+    void this.studioManagerPromise.then((studioManager) => {
```

Comment:
I'm wondering if it would be cleaner to set the studio manager object on the lifecycle manager once it's fully loaded, so we don't have to wait the promise in different places when we need it. We'd just have it.

Reply:
Yeah I think we'd still have the external wait on the promise, but internally, you'd just use the cached version if you know it's there.

Reply:
ah yeah I see what you mean. Updated 👍🏻